[SPARK-37833][INFRA] Add `precondition` job to skip the main GitHub Action jobs

dongjoon-hyun opened the pull request:
### What changes were proposed in this pull request?

This PR aims to introduce `precondition` jobs to skip the main GitHub Action jobs.

### Why are the changes needed?

This will save huge community GitHub Action resource and speed up our develop and PR review process.
- For example, GitHub Action will run only `linter` job for `docs` only PR.

<img width="468" alt="Screen Shot 2022-01-07 at 12 10 50 AM" src="https://user-images.githubusercontent.com/9700541/148512753-bd9b7e49-0e7b-47dd-9ce5-31f684dac666.png">


### Does this PR introduce _any_ user-facing change?

No. This is a dev-only infra.

### How was this patch tested?

Manually review and check the result on this PR.

dongjoon-hyun commented:
Hi, @HyukjinKwon .
Could you review this PR, please?

HyukjinKwon commented on .github/workflows/build_and_test.yml:
```diff
@@ -96,15 +96,39 @@ jobs:
+    - name: Check all modules except 'docs'
+      id: set-outputs
+      run: |
+        echo "::set-output name=required::$(./dev/is-changed.py -m avro,build,catalyst,core,docker-integration-tests,examples,graphx,hadoop-cloud,hive,hive-thriftserver,kubernetes,kvstore,launcher,mesos,mllib,mllib-local,network-common,network-shuffle,pyspark-core,pyspark-ml,pyspark-mllib,pyspark-pandas,pyspark-pandas-slow,pyspark-resource,pyspark-sql,pyspark-streaming,repl,sketch,spark-ganglia-l ...
```
One way to deduplicate this precondition job might be to execute this multiple times for the jobs of `build`, `pyspark`, `sparkr`, etc, and save the output as a JSON object such as:

```
echo '::set-output name=required::{"build": "yes", "pyspark": "no"}'
```

And then, we can use the JSON in the downstream job (e.g., `build`) as below:

```yaml
if fromJson(needs.build-precondition.outputs).build == 'yes'
```

e.g., https://github.com/apache/spark/blob/master/.github/workflows/build_and_test.yml#L83 and https://github.com/apache/spark/blob/master/.github/workflows/build_and_test.yml#L226

It's a bit sad that GItHub Actions doesn't have a feature like exit-early .. then the change would have been very simple .. 😢

dongjoon-hyun replied:
Yes, it's really a missing part 

BTW, `is-changed.py` is not designed to support multi-support. Each precondition has different `-m` option here.

dongjoon-hyun replied:
BTW, I'm planning to use `is-changed.py` in new M1 Jenkins too.

dongjoon-hyun committed:
[SPARK-37833][INFRA] Add precondition job to skip the main GitHub Action jobs

dongjoon-hyun commented:
Thank you, @HyukjinKwon . I addressed your comments. Could you review this once more, @HyukjinKwon ?
- `precondition` job runs all precondition checks and build `json` file.
- Each jobs `needs` only `precondition`'s result.

HyukjinKwon reviewed (approved):
LGTM

HyukjinKwon commented:
Oops, the test is broken in the commits (PRs are not affected). I made a quick followup but seems like we need some more fixes in `is-changed.py` script .. let me revert this and my own followup for now.
